File: 0649-dota2-senate/0649-dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r_queue = []
        d_queue = []
        n = len(senate)        

        for i, senator in enumerate(senate):
            if senator == 'R':
                r_queue.append(i)
            else:
                d_queue.append(i)

        while r_queue and d_queue:
            r_index = r_queue.pop(0)
            d_index = d_queue.pop(0)

            if r_index < d_index:
                r_queue.append(r_index + n)
            else:
                d_queue.append(d_index + n)

        return "Radiant" if r_queue else "Dire"
        

File: 0649-dota2-senate/0649-dota2-senate.rs
impl Solution {
    pub fn predict_party_victory(senate: String) -> String {
        let mut r_queue = std::collections::VecDeque::new();
        let mut d_queue = std::collections::VecDeque::new();
        let n = senate.len();        

        for (i, senator) in senate.chars().enumerate() {
            if senator == 'R' {
                r_queue.push_back(i);
            } else {
                d_queue.push_back(i);
            }
        }

        while !r_queue.is_empty() && !d_queue.is_empty() {
            let r_index = r_queue.pop_front().unwrap();
            let d_index = d_queue.pop_front().unwrap();

            if r_index < d_index {
                r_queue.push_back(r_index + n);
            } else {
                d_queue.push_back(d_index + n);
            }
        }

        if r_queue.is_empty() {
            "Dire".to_string()
        } else {
            "Radiant".to_string()
        }
    }
}